compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs

```diff
@@ -127,32 +116,69 @@ impl<I: Interner> NestedGoals<I> {
     }
 }
 
-#[derive(PartialEq, Eq, Debug, Hash, HashStable, Clone, Copy)]
+#[derive(PartialEq, Eq, Debug, Hash, HashStable_NoContext, Clone, Copy)]
 pub enum GenerateProofTree {
     Yes,
     No,
 }
 
-#[extension(pub trait InferCtxtEvalExt<'tcx>)]
-impl<'tcx> InferCtxt<'tcx> {
+pub trait SolverDelegateEvalExt: SolverDelegate {
+    fn evaluate_root_goal(
+        &self,
+        goal: Goal<Self::Interner, <Self::Interner as Interner>::Predicate>,
+        generate_proof_tree: GenerateProofTree,
+    ) -> (Result<(bool, Certainty), NoSolution>, Option<inspect::GoalEvaluation<Self::Interner>>);
+
+    fn evaluate_root_goal_raw(
```

Comment:
We have to expose this for use in `analyse.rs` which is still in `rustc_trait_selection`. Why? Because `analyse.rs` uses fulfillment (the `constrain` function uses an `ocx`).

 I could put a FIXME to remove this when fulfillment gets uplifted, though.